packages/template-retail-react-app/app/components/bonus-product-modal/index.jsx

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2025, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React, {useState, useEffect} from 'react'
+import {
+    Modal,
+    ModalCloseButton,
+    ModalContent,
+    ModalOverlay,
+    ModalFooter,
+    ModalHeader,
+    ModalBody,
+    Text,
+    Box,
+    VStack,
+    AspectRatio,
+    Skeleton,
+    SimpleGrid,
+    Button,
+    Checkbox
+} from '@salesforce/retail-react-app/app/components/shared/ui'
+import {useProducts} from '@salesforce/commerce-sdk-react'
+import DynamicImage from '@salesforce/retail-react-app/app/components/dynamic-image'
+import PropTypes from 'prop-types'
+import {useBonusProductModalContext} from '@salesforce/retail-react-app/app/hooks/use-bonus-product-modal'
+
+// Component to display individual bonus product with checkbox for selection
+const BonusProductItem = ({product, isSelected, onToggle, isLoading}) => {
+    const productName = product?.productName || product?.title || 'Unknown Product'
+    const productId = product?.productId || product?.id
+
+    // Fetch product data to get image information
+    const {data: productData, isLoading: isProductLoading} = useProducts(
+        {
+            parameters: {
+                ids: productId,
+                allImages: true
+            }
+        },
+        {
+            enabled: !!productId
+        }
+    )
+
+    // Extract image URL from imageGroups
+    const getProductImage = () => {
+        if (!productData?.data?.[0]?.imageGroups) return null
```

Comment:
Add a check for imageGroups to be an array or else it might fail on find function
```suggestion
        if (!productData?.data?.[0]?.imageGroups && !Array.isArray(productData.data[0].imageGroups)) return null
```